src/riscv/lib/src/machine_state.rs

```diff
@@ -328,10 +330,24 @@ impl<MC: memory::MemoryConfig, BCC: BlockCacheConfig, B: Block<MC, M>, M: backen
     where
         M: backend::ManagerReadWrite,
     {
-        self.core.reset();
+        self.core.reset(NoopMemoryGovernanceListener);
         self.block_cache.reset();
     }
 
+    /// Get access to the main memory, with a listener to hook into any permission updates.
+    ///
+    /// This is required to keep parts of [`MachineState`], that do not form part of the PVM state,
+    /// synchronised with main memory so as to ensure determinism regardless of whether these
+    /// additional parts of the state are populated or not.
+    pub(crate) fn memory_with_listener(
+        &mut self,
+    ) -> (
+        &mut MC::State<M>,
+        impl MemoryGovernanceListener + use<MC, BCC, B, M>,
+    ) {
+        (&mut self.core.main_memory, NoopMemoryGovernanceListener)
```

Comment:
If the returned listener is a no-op, why do you need a dedicated function for it? You can just instantiate a `NoopMemoryGovernanceListener` wherever needed.

Otherwise, at call-site I expect the listener to actually do something at least in _some_ cases. What do you say about having a constant exposed `DEFAULT_LISTENER/NOOP_LISTENER`? Avoids some noise destructuring the tuple at the same time & makes the behaviour more explicit at call site.

Reply:
because when we subsequently introduce the page cache, this will be replaced by the page cache here.

doing it in this order allows to update everywhere that needs to handle this, without making the page cache MR even more noisy

Reply:
A, ok, for the future change it makes sense 👍